Problem4/src/problem4/BouncingBallPanel.java

```diff
@@ -0,0 +1,176 @@
+        for (int i = 0; i < balls.length; i++) //Checks if each ball is outside of the frame (in other words, if it is colliding with the wall)
+        {
+            if (balls[i].x < balls[i].radius) //Left Wall
+            {
+                balls[i].dx = Math.abs(balls[i].dx);
+            }
+            if (balls[i].x > getWidth() - balls[i].radius) //Right Wall
+            {
+                balls[i].dx = -Math.abs(balls[i].dx);
+            }
+            if (balls[i].y < balls[i].radius) //Top Wall
+            {
+                balls[i].dy = Math.abs(balls[i].dy);
+            }
+            if (balls[i].y > getHeight() - balls[i].radius) //Bottom Wall
+            {
+                balls[i].dy = -Math.abs(balls[i].dy);
+            }
+
+            balls[i].x += balls[i].dx; //Updates the ball's x location based on speed
+            balls[i].y += balls[i].dy; //Updates the ball's y location based on speed
+            g.fillOval(balls[i].x - balls[i].radius, balls[i].y - balls[i].radius, balls[i].radius * 2, balls[i].radius * 2); //Draws the ball based on location
+        }
+    }
+    
+    /**
+     * Checks if any balls collide with each other and changes their speed accordingly
+     * 
+     * @param balls The array of balls
+     * @return The array of balls (after updating based on collision)
+     */
+    public Ball[] BallOnBallCollision(Ball[] balls)
+    {
+        double distance, xDist, yDist; //Declares variables for the overall, x, and y distances apart between two balls
+        Ball[] collidingBalls = new Ball[2]; //An array of two balls that have collided
+        for (int i = 0; i < 9; i++) //For each ball (except the last one)...
+        {
+            for (int j = i+1; j < 10; j++) //...and every ball after the one at index i
+            {
+                xDist = Math.abs(balls[i].x - balls[j].x); //Calculates the horizontal distance between the two balls
+                yDist = Math.abs(balls[i].y - balls[j].y); //Calculates the vertical distance between the two balls
+                distance = Math.sqrt(Math.pow(xDist, 2) + Math.pow(yDist, 2)); //Calculates the overall distance between the two balls
+                if (distance < balls[i].radius + balls[j].radius) //If the distance is less then the sum of the radiuses of the balls, they have collided
+                {
+                    collidingBalls = onCollide(balls[i], balls[j]); //Updates the balls' speed based on their collision
+                    balls[i] = collidingBalls[0];
+                    balls[j] = collidingBalls[1];
+                }
+            }
+        }
+        return balls;
+    }
+    
+    /**
+     * Updates the balls' speed based on their collision
+     * 
+     * @param a First ball
+     * @param b Second ball
+     * @return An array with both balls after the updates
+     */
+    public Ball[] onCollide(Ball a, Ball b) //Note: The algorithms used are not perfect
```

Comment:
Good job on your collision code.  However, your onCollide method should prevent a ball from stopping (I.e. movement should be non-zero in at least one direction).  There were a few times when a ball would bounce off another ball and then come to a stop.